diff --git a/fs/namei.c b/fs/namei.c
index 420734801..74859bb89 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -26,6 +26,10 @@
 #include <linux/sched/mm.h>
 #include <linux/fsnotify.h>
 #include <linux/personality.h>
+#ifdef CONFIG_VPNHIDE
+#include "../security/vpnhide/vpnhide.h"
+#endif
+
 #include <linux/security.h>
 #include <linux/ima.h>
 #include <linux/syscalls.h>
@@ -2618,6 +2622,11 @@ int filename_lookup(int dfd, struct filename *name, unsigned flags,
 		audit_inode(name, path->dentry,
 			    flags & LOOKUP_MOUNTPOINT ? AUDIT_INODE_NOEVAL : 0);
 	restore_nameidata();
+
+#ifdef CONFIG_VPNHIDE
+	if (!retval)
+		vpnhide_filename_lookup(dfd, name, flags, path, &retval);
+#endif
 	return retval;
 }
 
