diff --git a/kernel/bpf/syscall.c b/kernel/bpf/syscall.c
index d134efd31..7008524f5 100644
--- a/kernel/bpf/syscall.c
+++ b/kernel/bpf/syscall.c
@@ -6,6 +6,10 @@
 #include <linux/bpf_trace.h>
 #include <linux/bpf_lirc.h>
 #include <linux/bpf_verifier.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <linux/bsearch.h>
 #include <linux/btf.h>
 #include <linux/syscalls.h>
@@ -1495,6 +1499,10 @@ static int map_lookup_elem(union bpf_attr *attr)
 	err = bpf_map_copy_value(map, key, value, attr->flags);
 	if (err)
 		goto free_value;
 
+#ifdef CONFIG_VPNHIDE
+	vpnhide_bpf_lookup_elem(map, key, value);
+#endif
+
 	err = -EFAULT;
 	if (copy_to_user(uvalue, value, value_size) != 0)
@@ -4977,6 +4984,13 @@ static int bpf_map_do_batch(const union bpf_attr *attr,
 		BPF_DO_BATCH(map->ops->map_update_batch, map, f.file, attr, uattr);
 	else
 		BPF_DO_BATCH(map->ops->map_delete_batch, map, attr, uattr);
+#ifdef CONFIG_VPNHIDE
+	if (!err && (cmd == BPF_MAP_LOOKUP_BATCH ||
+	             cmd == BPF_MAP_LOOKUP_AND_DELETE_BATCH))
+		vpnhide_bpf_lookup_batch(map, attr, uattr);
+#endif
+
+
 err_put:
 	if (has_write)
 		bpf_map_write_active_dec(map);
