diff --git a/net/core/fib_rules.c b/net/core/fib_rules.c
index 75282222e..34a296a94 100644
--- a/net/core/fib_rules.c
+++ b/net/core/fib_rules.c
@@ -13,6 +13,10 @@
 #include <net/net_namespace.h>
 #include <net/sock.h>
 #include <net/fib_rules.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <net/ip_tunnels.h>
 #include <linux/indirect_call_wrapper.h>
 
@@ -1012,6 +1016,23 @@ static int fib_nl_fill_rule(struct sk_buff *skb, struct fib_rule *rule,
 	struct nlmsghdr *nlh;
 	struct fib_rule_hdr *frh;
 
+#ifdef CONFIG_VPNHIDE
+	if (is_target_uid()) {
+		if ((rule->iifname[0] && vpnhide_should_hide_ifname(rule->iifname)) ||
+		    (rule->oifname[0] && vpnhide_should_hide_ifname(rule->oifname)))
+			return 0;
+		{
+			uid_t _s = from_kuid(&init_user_ns, rule->uid_range.start);
+			uid_t _e = from_kuid(&init_user_ns, rule->uid_range.end);
+			if (_e != (uid_t)~0 &&
+			    (_s >= 10000 || _e >= 10000 ||
+			     is_target_uid_val(_s) || is_target_uid_val(_e)) &&
+			    rule->table != 253 && rule->table != 254 &&
+			    rule->table != 255 && rule->table > 100)
+				return 0;
+		}
+	}
+#endif
 	nlh = nlmsg_put(skb, pid, seq, type, sizeof(*frh), flags);
 	if (nlh == NULL)
 		return -EMSGSIZE;
