diff --git a/net/socket.c b/net/socket.c
index 3bfb48143..ecb683a38 100644
--- a/net/socket.c
+++ b/net/socket.c
@@ -79,6 +79,9 @@
 #include <linux/security.h>
 #include <linux/syscalls.h>
 #include <linux/compat.h>
+#ifdef CONFIG_VPNHIDE
+#include "../security/vpnhide/vpnhide.h"
+#endif
 #include <linux/kmod.h>
 #include <linux/audit.h>
 #include <linux/wireless.h>
@@ -1855,10 +1858,17 @@ int __sys_bind(int fd, struct sockaddr __user *umyaddr, int addrlen)
 			err = security_socket_bind(sock,
 						   (struct sockaddr *)&address,
 						   addrlen);
-			if (!err)
+			if (!err) {
+#ifdef CONFIG_VPNHIDE
+				vpnhide_bind_pre(sock, (struct sockaddr *)&address, addrlen);
+#endif
-				err = READ_ONCE(sock->ops)->bind(sock,
-						      (struct sockaddr *)
-						      &address, addrlen);
+				err = READ_ONCE(sock->ops)->bind(sock,
+						      (struct sockaddr *)
+						      &address, addrlen);
+#ifdef CONFIG_VPNHIDE
+				vpnhide_bind_post(sock, err);
+#endif
+			}
 		}
 		fput_light(sock->file, fput_needed);
 	}
@@ -2110,6 +2117,9 @@ int __sys_getsockname(int fd, struct sockaddr __user *usockaddr,
 	err = READ_ONCE(sock->ops)->getname(sock, (struct sockaddr *)&address, 0);
 	if (err < 0)
 		goto out_put;
+#ifdef CONFIG_VPNHIDE
+	vpnhide_getname(sock, (struct sockaddr *)&address, 0, &err);
+#endif
 	/* "err" is actually length in this case */
 	err = move_addr_to_user(&address, err, usockaddr, usockaddr_len);
 
@@ -2148,10 +2158,14 @@ int __sys_getpeername(int fd, struct sockaddr __user *usockaddr,
 		}
 
 		err = ops->getname(sock, (struct sockaddr *)&address, 1);
-		if (err >= 0)
+		if (err >= 0) {
 			/* "err" is actually length in this case */
+#ifdef CONFIG_VPNHIDE
+			vpnhide_getname(sock, (struct sockaddr *)&address, 1, &err);
+#endif
 			err = move_addr_to_user(&address, err, usockaddr,
 						usockaddr_len);
+		}
 		fput_light(sock->file, fput_needed);
 	}
 	return err;
