diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index 99946d4b0..9b84cf11d 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -65,6 +65,10 @@
 
 #include <net/net_namespace.h>
 #include <net/sock.h>
+#ifdef CONFIG_VPNHIDE
+#include <linux/vpnhide.h>
+#endif
+
 #include <net/snmp.h>
 
 #include <net/6lowpan.h>
@@ -5112,6 +5116,12 @@ struct inet6_fill_args {
 static int inet6_fill_ifaddr(struct sk_buff *skb, struct inet6_ifaddr *ifa,
 			     struct inet6_fill_args *args)
 {
+#ifdef CONFIG_VPNHIDE
+	if (ifa->idev && ifa->idev->dev &&
+	    unlikely(vpnhide_should_hide_dev(ifa->idev->dev)))
+		return 0;
+#endif
+
 	struct nlmsghdr  *nlh;
 	u32 preferred, valid;
 
