diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index 99946d4b0..9b84cf11d 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -65,6 +65,10 @@
 
 #include <net/net_namespace.h>
 #include <net/sock.h>
+#ifdef CONFIG_VPNHIDE
+#include <linux/vpnhide.h>
+#endif
+
 #include <net/snmp.h>
 
 #include <net/6lowpan.h>
@@ -3800,6 +3804,11 @@ static void if6_seq_stop(struct seq_file *seq, void *v)
 static int if6_seq_show(struct seq_file *seq, void *v)
 {
 	struct inet6_ifaddr *ifp = (struct inet6_ifaddr *)v;
+#ifdef CONFIG_VPNHIDE
+	if (ifp->idev && ifp->idev->dev &&
+	    vpnhide_should_hide_dev(ifp->idev->dev))
+		return 0;
+#endif
 	seq_printf(seq, "%pi6 %02x %02x %02x %02x %8s\n",
 		   &ifp->addr,
 		   ifp->idev->dev->ifindex,
