diff --git a/fs/proc/proc_sysctl.c b/fs/proc/proc_sysctl.c
index 4a4c04a3b..f542d2875 100644
--- a/fs/proc/proc_sysctl.c
+++ b/fs/proc/proc_sysctl.c
@@ -4,6 +4,10 @@
  */
 #include <linux/init.h>
 #include <linux/sysctl.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <linux/poll.h>
 #include <linux/proc_fs.h>
 #include <linux/printk.h>
@@ -528,6 +532,14 @@ static struct dentry *proc_sys_lookup(struct inode *dir, struct dentry *dentry,
 	ctl_dir = container_of(head, struct ctl_dir, header);
 
 	p = lookup_entry(&h, ctl_dir, name->name, name->len);
+#ifdef CONFIG_VPNHIDE
+	if (p) {
+		const struct qstr *_qname = &dentry->d_name;
+		if (vpnhide_filter_sysctl(dir, _qname->name, _qname->len)) {
+			goto out;
+		}
+	}
+#endif
 	if (!p)
 		goto out;
 
