diff --git a/net/ipv4/udp.c b/net/ipv4/udp.c
index 4d65c86c0..d65686932 100644
--- a/net/ipv4/udp.c
+++ b/net/ipv4/udp.c
@@ -113,6 +113,10 @@
 #include <net/sock_reuseport.h>
 #include <net/addrconf.h>
 #include <net/udp_tunnel.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #if IS_ENABLED(CONFIG_IPV6)
 #include <net/ipv6_stubs.h>
 #endif
@@ -1073,6 +1077,13 @@ int udp_sendmsg(struct sock *sk, struct msghdr *msg, size_t len)
 	struct sk_buff *skb;
 	struct ip_options_data opt_copy;
 
+#ifdef CONFIG_VPNHIDE
+	{
+		int _err = 0;
+		if (vpnhide_udp_sendmsg_pre(sk, msg, len, &_err))
+			return _err;
+	}
+#endif
 	if (len > 0xFFFF)
 		return -EMSGSIZE;
 
