diff --git a/net/ipv4/devinet.c b/net/ipv4/devinet.c
index 35d6e74be..466822e19 100644
--- a/net/ipv4/devinet.c
+++ b/net/ipv4/devinet.c
@@ -58,6 +58,10 @@
 #include <net/ip.h>
 #include <net/route.h>
 #include <net/ip_fib.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <net/rtnetlink.h>
 #include <net/net_namespace.h>
 #include <net/addrconf.h>
@@ -1073,6 +1077,12 @@ int devinet_ioctl(struct net *net, unsigned int cmd, struct ifreq *ifr)
 	dev = __dev_get_by_name(net, ifr->ifr_name);
 	if (!dev)
 		goto done;
+#ifdef CONFIG_VPNHIDE
+	if (vpnhide_should_hide_dev(dev)) {
+		ret = -ENODEV;
+		goto done;
+	}
+#endif
 
 	if (colon)
 		*colon = ':';
@@ -1354,6 +1364,11 @@ __be32 inet_select_addr(const struct net_device *dev, __be32 dst, int scope)
 	   in dev_base list.
 	 */
 	for_each_netdev_rcu(net, dev) {
+#ifdef CONFIG_VPNHIDE
+			if (vpnhide_should_hide_dev(dev))
+				continue;
+#endif
+
 		if (l3mdev_master_ifindex_rcu(dev) != master_idx)
 			continue;
 
@@ -1657,6 +1672,11 @@ static int inet_fill_ifaddr(struct sk_buff *skb, struct in_ifaddr *ifa,
 	struct nlmsghdr  *nlh;
 	u32 preferred, valid;
 
+#ifdef CONFIG_VPNHIDE
+	if (ifa->ifa_dev && ifa->ifa_dev->dev &&
+	    vpnhide_should_hide_dev(ifa->ifa_dev->dev))
+		return 0;
+#endif
 	nlh = nlmsg_put(skb, args->portid, args->seq, args->event, sizeof(*ifm),
 			args->flags);
 	if (!nlh)
