From: vpnhide <vpnhide@intree>
Subject: [PATCH] net: udp, ipv6/udp — rate limiting and link-local block

diff --git a/net/ipv4/udp.c b/net/ipv4/udp.c
--- a/net/ipv4/udp.c
+++ b/net/ipv4/udp.c
@@ -30,6 +30,7 @@
 #include <linux/slab.h>
 #include <linux/errqueue.h>
 #include <linux/socket.h>
+#include <linux/vpnhide.h>
 #include <net/sock.h>
 #include <net/udp.h>
 #include <net/route.h>
@@ -1147,6 +1150,10 @@ int udp_sendmsg(struct sock *sk, struct msghdr *msg, size_t len)
 	struct flowi4 fl4_stack;
 	int err;

+	if (unlikely(vpnhide_udp_sendmsg(sk))) {
+		err = -ENOBUFS;
+		goto out;
+	}
+
 	if (len > 0xFFFF)
 		return -EMSGSIZE;

diff --git a/net/ipv6/udp.c b/net/ipv6/udp.c
--- a/net/ipv6/udp.c
+++ b/net/ipv6/udp.c
@@ -30,6 +30,7 @@
 #include <linux/slab.h>
 #include <linux/errqueue.h>
 #include <linux/socket.h>
+#include <linux/vpnhide.h>
 #include <net/sock.h>
 #include <net/udp.h>
 #include <net/ipv6.h>
@@ -1460,6 +1463,10 @@ int udpv6_sendmsg(struct sock *sk, struct msghdr *msg, size_t len)
 	struct flowi6 fl6;
 	int err;

+	if (unlikely(vpnhide_udpv6_sendmsg_ll(sk, msg))) {
+		err = -ENODEV;
+		goto out;
+	}
+
 	if (len > INT_MAX)
 		return -EMSGSIZE;
