diff --git a/net/core/dev_ioctl.c b/net/core/dev_ioctl.c
index 0e87237fd..c34e022f9 100644
--- a/net/core/dev_ioctl.c
+++ b/net/core/dev_ioctl.c
@@ -9,6 +9,10 @@
 #include <linux/if_bridge.h>
 #include <net/dsa.h>
 #include <net/wext.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 
 #include "dev.h"
 
@@ -24,5 +28,11 @@ static int dev_ifname(struct net *net, struct ifreq *ifr)
 {
+	int _ret;
 	ifr->ifr_name[IFNAMSIZ-1] = 0;
-	return netdev_get_name(net, ifr->ifr_name, ifr->ifr_ifindex);
+	_ret = netdev_get_name(net, ifr->ifr_name, ifr->ifr_ifindex);
+#ifdef CONFIG_VPNHIDE
+	if (_ret == 0 && vpnhide_should_hide_ifname(ifr->ifr_name))
+		return -ENODEV;
+#endif
+	return _ret;
 }
 
@@ -63,6 +73,11 @@ int dev_ifconf(struct net *net, struct ifconf __user *uifc)
 	/* Loop over the interfaces, and write an info block for each. */
 	rtnl_lock();
 	for_each_netdev(net, dev) {
+
+#ifdef CONFIG_VPNHIDE
+		if (vpnhide_should_hide_dev(dev))
+			continue;
+#endif
 		if (!pos)
 			done = inet_gifconf(dev, NULL, 0, size);
 		else
@@ -139,6 +154,10 @@ static int dev_ifsioc_locked(struct net *net, struct ifreq *ifr, unsigned int cm
 
 	if (!dev)
 		return -ENODEV;
+#ifdef CONFIG_VPNHIDE
+	if (vpnhide_should_hide_dev(dev))
+		return -ENODEV;
+#endif
 
 	switch (cmd) {
 	case SIOCGIFFLAGS:	/* Get interface flags */
