diff --git a/net/ipv6/udp.c b/net/ipv6/udp.c
index 6e8efb688dda..31c8cdac6bce 100644
--- a/net/ipv6/udp.c
+++ b/net/ipv6/udp.c
@@ -57,6 +57,10 @@
 #include <linux/seq_file.h>
 #include <trace/events/skb.h>
 #include "udp_impl.h"
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <trace/hooks/net.h>
 
 static void udpv6_destruct_sock(struct sock *sk)
@@ -1504,6 +1508,13 @@ int udpv6_sendmsg(struct sock *sk, struct msghdr *msg, size_t len)
 
 	trace_android_rvh_udpv6_sendmsg(sk, msg, len);
 
+#ifdef CONFIG_VPNHIDE
+	{
+		int _err = 0;
+		if (vpnhide_udp_sendmsg_pre(sk, msg, len, &_err))
+			return _err;
+	}
+#endif
	ipcm6_init_sk(&ipc6, sk);
	ipc6.gso_size = READ_ONCE(up->gso_size);
