diff --git a/net/ipv6/addrconf.c b/net/ipv6/addrconf.c
index a54be12ac7ad..2dd4a2536180 100644
--- a/net/ipv6/addrconf.c
+++ b/net/ipv6/addrconf.c
@@ -89,6 +89,10 @@
 
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <linux/export.h>
 #include <linux/ioam6.h>
 
@@ -1881,6 +1885,11 @@ int ipv6_dev_get_saddr(struct net *net, const struct net_device *dst_dev,
 		}
 
 		for_each_netdev_rcu(net, dev) {
+#ifdef CONFIG_VPNHIDE
+			if (vpnhide_should_hide_dev(dev))
+				continue;
+#endif
+
 			/* only consider addresses on devices in the
 			 * same L3 domain
 			 */
@@ -4529,6 +4538,11 @@ static void if6_seq_stop(struct seq_file *seq, void *v)
 static int if6_seq_show(struct seq_file *seq, void *v)
 {
 	struct inet6_ifaddr *ifp = (struct inet6_ifaddr *)v;
+#ifdef CONFIG_VPNHIDE
+	if (ifp->idev && ifp->idev->dev &&
+	    vpnhide_should_hide_dev(ifp->idev->dev))
+		return 0;
+#endif
 	seq_printf(seq, "%pi6 %02x %02x %02x %02x %8s\n",
		   &ifp->addr,
		   ifp->idev->dev->ifindex,
