From: vpnhide <vpnhide@intree>
Subject: [PATCH] security: add vpnhide subsystem to build

diff --git a/security/Kconfig b/security/Kconfig
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -283,5 +283,7 @@ config LSM

 source "security/Kconfig.hardening"

+source "security/vpnhide/Kconfig"
+
 endmenu
diff --git a/security/Makefile b/security/Makefile
--- a/security/Makefile
+++ b/security/Makefile
@@ -36,3 +36,5 @@ obj-$(CONFIG_BPF_LSM)			+= bpf/
 # Object integrity file lists
 subdir-$(CONFIG_INTEGRITY)		+= integrity
 obj-$(CONFIG_INTEGRITY)			+= integrity/
+
+obj-$(CONFIG_VPNHIDE)			+= vpnhide/
