diff --git a/fs/namei.c b/fs/namei.c
index 0881626cd..ca10af316 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -24,6 +24,10 @@
 #include <linux/pagemap.h>
 #include <linux/fsnotify.h>
 #include <linux/personality.h>
+#ifdef CONFIG_VPNHIDE
+#include "../security/vpnhide/vpnhide.h"
+#endif
+
 #include <linux/security.h>
 #include <linux/ima.h>
 #include <linux/syscalls.h>
@@ -2579,6 +2583,11 @@ int filename_lookup(int dfd, struct filename *name, unsigned flags,
 			    flags & LOOKUP_MOUNTPOINT ? AUDIT_INODE_NOEVAL : 0);
 	restore_nameidata();
+
+#ifdef CONFIG_VPNHIDE
+	if (!retval)
+		vpnhide_filename_lookup(dfd, name, flags, path, &retval);
+#endif
 	putname(name);
 	return retval;
 }
 
