diff --git a/kernel/bpf/syscall.c b/kernel/bpf/syscall.c
index 589332936..39368762d 100644
--- a/kernel/bpf/syscall.c
+++ b/kernel/bpf/syscall.c
@@ -5,6 +5,10 @@
 #include <linux/bpf_trace.h>
 #include <linux/bpf_lirc.h>
 #include <linux/bpf_verifier.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 #include <linux/btf.h>
 #include <linux/syscalls.h>
 #include <linux/slab.h>
@@ -1059,6 +1063,9 @@ static int map_lookup_elem(union bpf_attr *attr)
 		goto free_key;
 
 	err = bpf_map_copy_value(map, key, value, attr->flags);
+#ifdef CONFIG_VPNHIDE
+	vpnhide_bpf_lookup_elem(map, key, value);
+#endif
 	if (err)
 		goto free_value;
 
@@ -4032,5 +4039,10 @@ static int bpf_map_do_batch(const union bpf_attr *attr,
 		BPF_DO_BATCH(map->ops->map_update_batch);
 	else
 		BPF_DO_BATCH(map->ops->map_delete_batch);
+#ifdef CONFIG_VPNHIDE
+	if (!err && (cmd == BPF_MAP_LOOKUP_BATCH ||
+	             cmd == BPF_MAP_LOOKUP_AND_DELETE_BATCH))
+		vpnhide_bpf_lookup_batch(map, attr, uattr);
+#endif
 err_put:
