diff --git a/net/core/dev_ioctl.c b/net/core/dev_ioctl.c
index 993420da2..8e318e847 100644
--- a/net/core/dev_ioctl.c
+++ b/net/core/dev_ioctl.c
@@ -8,6 +8,9 @@
 #include <linux/wireless.h>
 #include <net/dsa.h>
 #include <net/wext.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
 
 /*
  *	Map an interface index to its name (SIOCGIFNAME)
@@ -22,8 +25,14 @@
 
 static int dev_ifname(struct net *net, struct ifreq *ifr)
 {
+	int _ret;
 	ifr->ifr_name[IFNAMSIZ-1] = 0;
-	return netdev_get_name(net, ifr->ifr_name, ifr->ifr_ifindex);
+	_ret = netdev_get_name(net, ifr->ifr_name, ifr->ifr_ifindex);
+#ifdef CONFIG_VPNHIDE
+	if (_ret == 0 && vpnhide_should_hide_ifname(ifr->ifr_name))
+		return -ENODEV;
+#endif
+	return _ret;
 }
 
 /*
@@ -85,6 +94,11 @@ static int dev_ifsioc_locked(struct net *net, struct ifreq *ifr, unsigned int cm
 	if (!dev)
 		return -ENODEV;
 
+#ifdef CONFIG_VPNHIDE
+	if (vpnhide_should_hide_dev(dev))
+		return -ENODEV;
+#endif
+
 	switch (cmd) {
 	case SIOCGIFFLAGS:	/* Get interface flags */
 		ifr->ifr_flags = (short) dev_get_flags(dev);
