diff --git a/net/ipv4/udp.c b/net/ipv4/udp.c
index e1284a622..d62f5d606 100644
--- a/net/ipv4/udp.c
+++ b/net/ipv4/udp.c
@@ -117,6 +117,10 @@
 #include <net/ipv6_stubs.h>
 #endif
 #include <trace/hooks/ipv4.h>
+#ifdef CONFIG_VPNHIDE
+#include "../../security/vpnhide/vpnhide.h"
+#endif
+
 
 struct udp_table udp_table __read_mostly;
 EXPORT_SYMBOL(udp_table);
@@ -1058,6 +1062,13 @@ int udp_sendmsg(struct sock *sk, struct msghdr *msg, size_t len)
 	struct sk_buff *skb;
 	struct ip_options_data opt_copy;
 
+#ifdef CONFIG_VPNHIDE
+	{
+		int _err = 0;
+		if (vpnhide_udp_sendmsg_pre(sk, msg, len, &_err))
+			return _err;
+	}
+#endif
 	if (len > 0xFFFF)
 		return -EMSGSIZE;
 
